posthog/hogql/printer.py

```diff
@@ -433,9 +439,17 @@ def visit_compare_operation(self, node: ast.CompareOperation):
             op = f"notILike({left}, {right})"
             value_if_one_side_is_null = True
         elif node.op == ast.CompareOperationOp.In:
-            op = f"in({left}, {right})"
+            operator = "in"
+            # External tables need to use globalIn with distributed tables (events)
+            if left_is_events and right_is_s3:
+                operator = "globalIn"
+            op = f"{operator}({left}, {right})"
         elif node.op == ast.CompareOperationOp.NotIn:
-            op = f"notIn({left}, {right})"
+            operator = "notIn"
+            # External tables need to use globalNotIn with distributed tables (events)
+            if left_is_events and right_is_s3:
+                operator = "globalNotIn"
```

Comment:
I would say this is not the right level to make this call at. The `Printer` needs to be relatively dumb, taking a fully resolved AST and naively converting it to a string. This is better suited for the `Resolver`, which is where we can swap `ast.CompareOperationOp.In` to `ast.CompareOperationOp.GlobalIn`

Reply:
agreed, moved